pandas/_libs/internals.pyx

```diff
@@ -890,29 +890,29 @@ cdef class BlockValuesRefs:
     """
     cdef:
         public list referenced_blocks
-        public int clear_counter
+        public int dead_counter
+        object __weakref__
+        object _weakref_cb
 
     def __cinit__(self, blk: Block | None = None) -> None:
+        def _weakref_cb(item: weakref.ref, selfref: weakref.ref = weakref.ref(self)) -> None:
```

Comment:
Not sure of all of the impacts but I think `cinit` is mostly used for C-level structures. I think setting this callback should instead be done in `init`

Reply:
I just updated it

Reply:
Actually, it's doesn't work because __cinit__ is called before __init__, so we need the callback ready there, unless we move everything to __init__. I changed it back 😅

Reply:
Hmm OK. And this definitely doesn't cause any memory leaks or issues with reference counting then right?

Reply:
It should be correct as long as the reference list is not modified manually without the classes methods or run under concurrency.

Another thing is that taking the length of the list of refs directly won't give the correct length too (the actual value is len(referenced_blocks) - dead_counter). From what I looked, the exact length was never used anywhere else in the code right now, but I can implement the len method if needed.

If someone modified the list manually without the classes methods, the counter would be wrong and the memory consumption could grow a lot (just like the current solution at main) and the has_reference() will be plain wrong.

Reply:
> run under concurrency.

Can you elaborate? I am pretty sure that pandas is used under concurrency where this might be problematic

Reply:
Sorry, I meant multi threading

Reply:
Yes I understood this, but again, this is a realistic scenario

Reply:
The problem would be that `self.dead_counter += 1` isn't atomic. I can add a lock to it but, correct if I'm wrong, the rest of the code isn't thread-safe too. e.g. list comprehension from _clear_dead_references doesn't seem to be thread-safe

Reply:
Here is a sample of how _clear_dead_references is not really thread safe:
```
import threading
class Test:
    def __init__(self):
        self.v = list(range(1000))

    def add(self, x):
        self.v.append(x)

    def rebuild(self):
        self.v = [x for x in self.v]
    
    def simulate(self):
        for _ in range(1000):
            self.add(1)
            self.rebuild()
    
    def race(self):
        threads = [threading.Thread(target=self.simulate) for _ in range(10)]
        for t in threads:
            t.start()
        for t in threads:
            t.join()
        print(len(self.v))

test1 = Test()
test2 = Test()
test1.race()
test2.race()
```

I couldn't reproduce the same with pandas code, but I think it's more about timing than something else. The point is, is this class supposed to be thread-safe?